frontend/src/pages/answer/fullKg/KgFull.jsx

```diff
@@ -0,0 +1,127 @@
+/* eslint-disable indent, no-use-before-define, func-names, no-return-assign */
+import React, {
+  useState, useContext, useEffect, useRef,
+} from 'react';
+import * as d3 from 'd3';
+import CircularProgress from '@material-ui/core/CircularProgress';
+import Box from '@material-ui/core/Box';
+
+import BiolinkContext from '~/context/biolink';
+import kgUtils from '~/utils/knowledgeGraph';
+
+import Worker from './simulation.worker';
+
+import './kgFull.css';
+
+const height = 400;
+const width = 400;
+// const maxNodes = 40;
+// const maxEdges = 100;
+
+/**
+ * Full Knowledge Graph display
+ * @param {object} message - TRAPI message
+ */
+export default function KgFull({ message }) {
+  const canvasRef = useRef();
+  const [loading, toggleLoading] = useState(false);
+  const [progress, setProgress] = useState(0);
+  const { colorMap, hierarchies } = useContext(BiolinkContext);
+
+  function displayCanvas(data) {
+    const canvas = d3.select(canvasRef.current)
+      .attr('width', width)
+      .attr('height', height);
+    const context = canvas.node().getContext('2d');
+    const { nodes, edges } = data;
+    context.clearRect(0, 0, width, height);
+    context.save();
+    context.translate(width / 2, height / 2);
+
+    function drawEdge(d) {
+      context.moveTo(d.source.x, d.source.y);
+      context.lineTo(d.target.x, d.target.y);
+    }
+
+    function drawNode(d) {
+      context.beginPath();
+      context.moveTo(d.x + 5, d.y);
+      context.arc(d.x, d.y, 5, 0, 2 * Math.PI);
+      if (d.category && Array.isArray(d.category)) {
+        d.category = kgUtils.getRankedCategories(hierarchies, d.category);
+      }
+      context.strokeStyle = colorMap((d.category && d.category[0]) || 'unknown');
+      context.fillStyle = colorMap((d.category && d.category[0]) || 'unknown');
+      context.fill();
+      context.stroke();
+    }
+
+    context.beginPath();
+    edges.forEach(drawEdge);
+    context.strokeStyle = '#aaa';
+    context.stroke();
+
+    nodes.forEach(drawNode);
+
+    context.restore();
+  }
+
+  useEffect(() => {
```

Comment:
I think it might help readability if big useEffect functions were not inlined, and were named.

Reply:
Fixed in b6653ab66a4515b3d14226e2f9311c169f3ea0d1